test/test_builders.go

```diff
@@ -0,0 +1,156 @@
+	}
+
+	packCmd := filepath.Join(wd, "pack")
+
+	funcBinaries := []string {
+		filepath.Join(wd, "bin", "func_snapshot"),
+		filepath.Join(wd, "bin", "func_stable"),
+	}
+	templates := []string {"http", "events" }
+
+	builders := []struct {
+		BuilderImage string
+		Runtime      []string
+	}{
+		{
+			BuilderImage: "quay.io/boson/faas-go-builder",
+			Runtime: []string{"go"},
+		},
+		{
+			BuilderImage: "quay.io/boson/faas-jvm-builder",
+			Runtime: []string{"quarkus", "springboot" },
+		},
+		{
+			BuilderImage: "quay.io/boson/faas-nodejs-builder",
+			Runtime: []string{"node", "typescript"},
+		},
+		{
+			BuilderImage: "quay.io/boson/faas-python-builder",
+			Runtime: []string{"python"},
+		},
+		{
+			BuilderImage: "quay.io/boson/faas-quarkus-native-builder",
+			Runtime: []string{"quarkus"},
+		},
+		{
+			BuilderImage: "quay.io/boson/faas-rust-builder",
+			Runtime: []string{"rust"},
+		},
+	}
+
+	for i := range builders {
+		builders[i].BuilderImage = fmt.Sprintf("%s:%s", builders[i].BuilderImage, version)
+	}
+
+	oldWD, err := os.Getwd()
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer os.Chdir(oldWD)
+
+	testDir, err := os.MkdirTemp("", "test_func")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer os.RemoveAll(testDir)
+	os.Chdir(testDir)
+
+	var errors []string
+	var fnNumber int
+	for _, funcBinary := range funcBinaries {
+		for _, builder := range builders {
+			for _, runtime := range builder.Runtime {
+				for _, template := range templates {
+
+					fnNumber++
+					func() {
+						defer os.Chdir(testDir)
+
+						fnName := fmt.Sprintf("fnno-%d", fnNumber)
+
+						tp := fmt.Sprintf("test_params{ binary: %q, runtime: %q, template: %q, builder %q }",
+							funcBinary, runtime, template, builder.BuilderImage)
+
+						log.Printf("building function %s", tp)
+
+						cmd := exec.Command(funcBinary,
+							"create", fnName,
+							"--runtime", runtime,
+							"--template", template,
+							"--verbose")
+						cmd.Stdout = os.Stdout
+						cmd.Stderr = os.Stderr
+
+						err := cmd.Run()
+						if err != nil {
+							msg := fmt.Sprintf("failed to create a function (%s): %s", tp, err.Error())
+							log.Println(msg)
+							errors = append(errors, msg)
+							return
+						}
+						os.Chdir(fnName)
+
+						cmd = exec.Command(packCmd,
+							"build", "example.com/jdoe/" + fnName + ":latest",
+							"--builder", builder.BuilderImage,
+							"--pull-policy=if-not-present",
+							"--trust-builder=false",
+							"--verbose")
+						cmd.Stdout = os.Stdout
+						cmd.Stderr = os.Stderr
+
+						err = cmd.Run()
+						if err != nil {
+							msg := fmt.Sprintf("failed to build the function using `pack` (--trust-builder=false, %s): %s", tp, err.Error())
+							log.Println(msg)
+							errors = append(errors, msg)
+						}
+
+						cmd = exec.Command(packCmd,
+							"build", "example.com/jdoe/" + fnName + ":latest",
+							"--builder", builder.BuilderImage,
+							"--pull-policy=if-not-present",
+							"--trust-builder=true",
```

Comment:
Why run both `--trust-builder=true` and `--trust-builder=false`? Don't we specify it's a trusted builder internally in `func`?

Reply:
we do I believe it should work with both

Reply:
Also if you wonder why building by `pack` and not by `func` -- `func` always pulls images so we wouldn't be testing what we want to

Reply:
> Also if you wonder why building by `pack` and not by `func` -- `func` always pulls images so we wouldn't be testing what we want to

I often test with `pack` locally for this reason. Another option, which I have been doing lately is to tag the new builder images with `tip-test` and those never get pushed, so I ensure I'm testing with my new local image. https://gist.github.com/lance/b54dd596564106e079530f55cce42548#file-test-builders-sh-L35

Reply:
@lance I am not sure if just tagging the builder images is enough, builders internally contains references to another images and those references won't be changed.

Reply:
> @lance I am not sure if just tagging the builder images is enough, builders internally contains references to another images and those references won't be changed.

I suppose it depends on what parts of the builder have been changed. This should work if I am changing/testing functionality in the build or detect scripts for example. But yes, that's a good point.